### PR TITLE
Tag RSPECs of JS/TS rules relying on type information

### PR DESCRIPTION
**Pull request by yassin-kammoun-sonarsource**
## Review

A dedicated reviewer checked the rule description successfully for:

- [ ] logical errors and incorrect information
- [ ] information gaps and missing content
- [ ] text style and tone
- [ ] PR summary and labels follow [the guidelines](https://github.com/SonarSource/rspec/#to-modify-an-existing-rule)

### CODE REVIEW

File: rules/S2871/javascript/metadata.json [hunk 1]
```diff
@@ -13,7 +13,8 @@
     "constantCost": "10min"
   },
   "tags": [
-    "bad-practice"
+    "bad-practice",
+    "type-dependent"
```
Comment (victor-diez-sonarsource) [inline rules/S2871/javascript/metadata.json:17]: +1

File: rules/S2999/javascript/metadata.json [hunk 2]
```diff
@@ -13,7 +13,7 @@
     "constantCost": "10min"
   },
   "tags": [
-
+    "type-dependent"
```
Comment (victor-diez-sonarsource) [inline rules/S2999/javascript/metadata.json:16]: +1

File: rules/S3402/javascript/metadata.json [hunk 3]
```diff
@@ -13,7 +13,8 @@
     "constantCost": "15min"
   },
   "tags": [
-    "confusing"
+    "confusing",
+    "type-dependent"
```
Comment (victor-diez-sonarsource) [inline rules/S3402/javascript/metadata.json:17]: +1

File: rules/S3403/javascript/metadata.json [hunk 4]
```diff
@@ -1,4 +1,7 @@
 {
   "title": "Strict equality operators should not be used with dissimilar types",
-  "quickfix": "covered"
+  "quickfix": "covered",
+  "tags": [
+    "type-dependent"
```
Comment (victor-diez-sonarsource) [inline rules/S3403/javascript/metadata.json:5]: +1

File: rules/S3525/javascript/metadata.json [hunk 5]
```diff
@@ -13,7 +13,8 @@
     "constantCost": "5min"
   },
   "tags": [
-    "es2015"
+    "es2015",
+    "type-dependent"
```
Comment (victor-diez-sonarsource) [inline rules/S3525/javascript/metadata.json:17]: it can also work without type information

File: rules/S3533/javascript/metadata.json [hunk 6]
```diff
@@ -15,7 +15,8 @@
   "tags": [
     "convention",
     "es2015",
-    "obsolete"
+    "obsolete",
+    "type-dependent"
```
Comment (victor-diez-sonarsource) [inline rules/S3533/javascript/metadata.json:19]: it can also work without type information

File: rules/S3579/javascript/metadata.json [hunk 7]
```diff
@@ -13,7 +13,8 @@
     "constantCost": "10min"
   },
   "tags": [
-    "bad-practice"
+    "bad-practice",
+    "type-dependent"
```
Comment (victor-diez-sonarsource) [inline rules/S3579/javascript/metadata.json:17]: +1

File: rules/S3735/javascript/metadata.json [hunk 8]
```diff
@@ -13,7 +13,8 @@
     "constantCost": "5min"
   },
   "tags": [
-    "confusing"
+    "confusing",
+    "type-dependent"
```
Comment (victor-diez-sonarsource) [inline rules/S3735/javascript/metadata.json:17]: it can also work without type information

File: rules/S3757/javascript/metadata.json [hunk 9]
```diff
@@ -13,7 +13,7 @@
     "constantCost": "10min"
   },
   "tags": [
-
+    "type-dependent"
```
Comment (victor-diez-sonarsource) [inline rules/S3757/javascript/metadata.json:16]: +1

File: rules/S3758/javascript/metadata.json [hunk 10]
```diff
@@ -13,7 +13,7 @@
     "constantCost": "10min"
   },
   "tags": [
-
+    "type-dependent"
```
Comment (victor-diez-sonarsource) [inline rules/S3758/javascript/metadata.json:16]: +1

File: rules/S3760/javascript/metadata.json [hunk 11]
```diff
@@ -13,7 +13,8 @@
     "constantCost": "2min"
   },
   "tags": [
-    "confusing"
+    "confusing",
+    "type-dependent"
```
Comment (victor-diez-sonarsource) [inline rules/S3760/javascript/metadata.json:17]: +1

File: rules/S3782/javascript/metadata.json [hunk 12]
```diff
@@ -13,7 +13,7 @@
     "constantCost": "5min"
   },
   "tags": [
-
+    "type-dependent"
```
Comment (victor-diez-sonarsource) [inline rules/S3782/javascript/metadata.json:16]: +1

File: rules/S3785/javascript/metadata.json [hunk 13]
```diff
@@ -13,7 +13,7 @@
     "constantCost": "5min"
   },
   "tags": [
-
+    "type-dependent"
```
Comment (victor-diez-sonarsource) [inline rules/S3785/javascript/metadata.json:16]: +1

File: rules/S3796/javascript/metadata.json [hunk 14]
```diff
@@ -13,7 +13,7 @@
     "constantCost": "5min"
   },
   "tags": [
-
+    "type-dependent"
```
Comment (victor-diez-sonarsource) [inline rules/S3796/javascript/metadata.json:16]: +1

File: rules/S3800/javascript/metadata.json [hunk 15]
```diff
@@ -13,7 +13,8 @@
     "constantCost": "20min"
   },
   "tags": [
-    "confusing"
+    "confusing",
+    "type-dependent"
```
Comment (victor-diez-sonarsource) [inline rules/S3800/javascript/metadata.json:17]: +1

File: rules/S4043/javascript/metadata.json [hunk 16]
```diff
@@ -13,7 +13,7 @@
     "constantCost": "5min"
   },
   "tags": [
-
+    "type-dependent"
```
Comment (victor-diez-sonarsource) [inline rules/S4043/javascript/metadata.json:16]: +1

File: rules/S4123/javascript/metadata.json [hunk 17]
```diff
@@ -13,7 +13,8 @@
     "constantCost": "1min"
   },
   "tags": [
-    "confusing"
+    "confusing",
+    "type-dependent"
```
Comment (victor-diez-sonarsource) [inline rules/S4123/javascript/metadata.json:17]: +1

File: rules/S4139/javascript/metadata.json [hunk 18]
```diff
@@ -13,7 +13,7 @@
     "constantCost": "5min"
   },
   "tags": [
-
+    "type-dependent"
```
Comment (victor-diez-sonarsource) [inline rules/S4139/javascript/metadata.json:16]: +1

File: rules/S4324/javascript/metadata.json [hunk 19]
```diff
@@ -13,7 +13,8 @@
     "constantCost": "2min"
   },
   "tags": [
-    "proficiency"
+    "proficiency",
+    "type-dependent"
```
Comment (victor-diez-sonarsource) [inline rules/S4324/javascript/metadata.json:17]: +1

File: rules/S4335/javascript/metadata.json [hunk 20]
```diff
@@ -13,7 +13,8 @@
     "constantCost": "5min"
   },
   "tags": [
-    "pitfall"
+    "pitfall",
+    "type-dependent"
```
Comment (victor-diez-sonarsource) [inline rules/S4335/javascript/metadata.json:17]: +1

File: rules/S4619/javascript/metadata.json [hunk 21]
```diff
@@ -13,7 +13,8 @@
     "constantCost": "5min"
   },
   "tags": [
-    "pitfall"
+    "pitfall",
+    "type-dependent"
```
Comment (victor-diez-sonarsource) [inline rules/S4619/javascript/metadata.json:17]: +1

File: rules/S4623/javascript/metadata.json [hunk 22]
```diff
@@ -13,7 +13,8 @@
     "constantCost": "1min"
   },
   "tags": [
-    "confusing"
+    "confusing",
+    "type-dependent"
```
Comment (victor-diez-sonarsource) [inline rules/S4623/javascript/metadata.json:17]: +1

File: rules/S4822/javascript/metadata.json [hunk 23]
```diff
@@ -13,7 +13,7 @@
     "constantCost": "5min"
   },
   "tags": [
-
+    "type-dependent"
```
Comment (victor-diez-sonarsource) [inline rules/S4822/javascript/metadata.json:16]: +1

File: rules/S5247/javascript/metadata.json [hunk 24]
```diff
@@ -1,5 +1,9 @@
 {
   "defaultQualityProfiles": [
     "Sonar way"
+  ],
+  "tags": [
+    "cwe",
+    "type-dependent"
```
Comment (victor-diez-sonarsource) [inline rules/S5247/javascript/metadata.json:7]: it can work without type information

File: rules/S5725/javascript/metadata.json [hunk 25]
```diff
@@ -2,5 +2,9 @@
   "defaultQualityProfiles": [
     "Sonar way"
   ],
-  "scope": "Main"
+  "scope": "Main",
+  "tags": [
+    "cwe",
+    "type-dependent"
```
Comment (victor-diez-sonarsource) [inline rules/S5725/javascript/metadata.json:8]: +1

File: rules/S5856/javascript/metadata.json [hunk 26]
```diff
@@ -1,3 +1,6 @@
 {
-  "scope": "Main"
+  "scope": "Main",
+  "tags": [
+    "type-dependent"
```
Comment (victor-diez-sonarsource) [inline rules/S5856/javascript/metadata.json:4]: it can also work without type information

File: rules/S6594/javascript/metadata.json [hunk 27]
```diff
@@ -13,7 +13,8 @@
     "constantCost": "5min"
   },
   "tags": [
-    "regex"
+    "regex",
+    "type-dependent"
```
Comment (victor-diez-sonarsource) [inline rules/S6594/javascript/metadata.json:17]: +1

File: rules/S6759/javascript/metadata.json [hunk 28]
```diff
@@ -7,7 +7,8 @@
     "constantCost": "5min"
   },
   "tags": [
-    "react"
+    "react",
+    "type-dependent"
```
Comment (victor-diez-sonarsource) [inline rules/S6759/javascript/metadata.json:11]: +1

**Comment by yassin-kammoun-sonarsource** [comment]
> still have external rules to check, but I saw some rules not 100% dependant on type information that would still work. `+1` means checked and agreed

If by external rules you mean rules from TypeScript ESLint or eslint-plugin-sonarjs, I already checked them. I also know that there are some rules that will partially work if type information is missing. The point here is to know which ones rely fully or partially on it so that it is easy for us to find them as well as users.

For the record, this tag will be mentioned in https://github.com/SonarSource/SonarJS/issues/4186 so that SonarLint users can know which rules might not behave as expected when type information is missing. For now, I don't think it's worth distinguishing partial from full dependence on type information.

**Review by victor-diez-sonarsource — APPROVED** [review]

**Pull request merged**